evm/src/arithmetic/byte.rs

```diff
@@ -0,0 +1,502 @@
+    // can be range-checked; but the rigidity of the range-checking
+    // mechanism means we can't optionally switch it off for some
+    // instructions.
+    lv[BYTE_IDX_HI_LIMB_SUM_INV_0] = F::from_canonical_u16(hi_limb_sum_inv as u16);
+    lv[BYTE_IDX_HI_LIMB_SUM_INV_1] = F::from_canonical_u16((hi_limb_sum_inv >> 16) as u16);
+    lv[BYTE_IDX_HI_LIMB_SUM_INV_2] = F::from_canonical_u16((hi_limb_sum_inv >> 32) as u16);
+    lv[BYTE_IDX_HI_LIMB_SUM_INV_3] = F::from_canonical_u16((hi_limb_sum_inv >> 48) as u16);
+    lv[BYTE_IDX_IS_LARGE] = F::from_bool(!hi_limb_sum.is_zero());
+
+    // Set the tree values according to the low 5 bits of idx, even
+    // when idx >= 32.
+
+    // Use the bits of idx0 to build a multiplexor that selects
+    // the correct byte of val. Each level of the tree uses one
+    // bit to halve the set of possible bytes from the previous
+    // level. The tree stores limbs rather than bytes though, so
+    // the last value must be handled specially.
+
+    // Morally, offset at i is 2^i * bit[i], but because of the
+    // reversed indexing and handling of the last element
+    // separately, the offset is 2^i * ( ! bit[i + 1]). (The !bit
+    // corresponds to calculating 31 - bits which is just bitwise NOT.)
+
+    // Conceptually we want to initialise the tree using something
+    // like this:
+    //
+    //   let tree = &mut lv[AUX_INPUT_REGISTER_1];
+    //   let val_limbs = &lv[INPUT_REGISTER_1];
+    //   tree[..8].copy_from_slice(&val_limbs[offset..offset + 8]);
+    //
+    // but we can't borrow both `tree` and `val_limbs` simultaneously.
+    // Apparently the solution is to use `split_at_mut()`; below we
+    // assume that the `val` registers are earlier in the row than the
+    // `tree` registers, so we enforce that assumption here.
+    let val_idx = INPUT_REGISTER_1.start;
+    let tree_idx = AUX_INPUT_REGISTER_1.start;
+    assert!(val_idx + N_LIMBS < tree_idx);
+
+    // `lvl_len` is the number of elements of the current level of the
+    // "tree". Can think of `val_limbs` as level 0, with length =
+    // N_LIMBS = 16.
+    const_assert!(N_LIMBS == 16); // Enforce assumption
+
+    let (prev, next) = lv[val_idx..].split_at_mut(tree_idx - val_idx);
+    let lvl_len = 8;
+    let offset = (!idx.bit(4) as usize) * lvl_len;
+    next[..lvl_len].copy_from_slice(&prev[offset..offset + lvl_len]);
+
+    let (prev, next) = next.split_at_mut(lvl_len);
+    let lvl_len = 4;
+    let offset = (!idx.bit(3) as usize) * lvl_len;
+    next[..lvl_len].copy_from_slice(&prev[offset..offset + lvl_len]);
+
+    let (prev, next) = next.split_at_mut(lvl_len);
+    let lvl_len = 2;
+    let offset = (!idx.bit(2) as usize) * lvl_len;
+    next[..lvl_len].copy_from_slice(&prev[offset..offset + lvl_len]);
+
+    let (prev, next) = next.split_at_mut(lvl_len);
+    // lvl_len = 1
+    let offset = !idx.bit(1) as usize;
+    next[0] = prev[offset]; // tree[14] = tree[offset + 12]
```

Comment:
How about [`copy_within`](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_within)?

Reply:
Haha, StackOverflow led me astray! Agreed, that's much cleaner.